Allow builders to override support for linkcode references

jdknight opened the pull request:
> Note: this was originally proposed in https://github.com/sphinx-doc/sphinx/pull/10585, but was accidentally removed during some cleanup; republishing.

### Feature or Bugfix

- Feature

### Purpose

The original `sphinx.ext.linkcode` extension will only include references into a doctree for `html` builders. Custom builders who would like to handle linkcode references do not have a graceful way to enable this feature.

This commit aims to allow custom builders to override when these references are included. If a builder defines a `supported_linkcode` attribute, this extension will use the value of this attribute to consider other builders/formats when conditionally adding references.

### Relates
- https://github.com/sphinx-contrib/confluencebuilder/issues/421
- https://github.com/sphinx-doc/sphinx/pull/10585

jdknight committed:
Allow builders to flag support for linkcode references

The original `sphinx.ext.linkcode` extension will only include
references into a doctree for `html` builders. Custom builders who
would like to handle linkcode references do not have a graceful way to
enable this feature.

This commit aims to allow custom builders to override when these
references are included. If a builder defines a `supported_linkcode`
attribute, this extension will use the value of this attribute to
consider other builders/formats when conditionally adding references.

Signed-off-by: James Knight <james.d.knight@live.com>

picnixz commented on doc/extdev/extensions/linkcode.rst:
```diff
@@ -0,0 +1,25 @@
+``sphinx.ext.linkcode`` API
+===========================
+
+Refer to :mod:`sphinx.ext.linkcode` more for information about the
```
```suggestion
Refer to :mod:`sphinx.ext.linkcode` for more information about the
```

picnixz commented on doc/extdev/extensions/linkcode.rst:
```diff
@@ -0,0 +1,25 @@
+      .. code-block:: python
+
+         class MyAwesomeBuilder(Builder):
+             supported_linkcode = True
```
Like this, you would have an only node with `True` and not 'html'.

picnixz reviewed (commented):
Some comments:

- ~I would expect `supports_linkcode = True` to indicate `'html'`~
- ~I would expect `supports_linkcode = False` to entirely disable linkcode support.~
- I would recommend the documentation to explain that whatever is in `supports_linkcode` should match the `only` directive argument.

jdknight committed:
Add extdev doc for interfacing with linkcode

Adding a document under Sphinx's "extension development" chapter which
provides information about integrating third-party builds with recently
added features to the built-in linkcode extension.

Signed-off-by: James Knight <james.d.knight@live.com>

AA-Turner committed:
Move docs
